pkg/interface/chat/src/js/components/lib/invite-search.js

```diff
@@ -0,0 +1,274 @@
+import React, { Component } from 'react';
+import urbitOb from "urbit-ob";
+import { Sigil } from "../lib/icons/sigil";
+
+
+export class InviteSearch extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      groups: [],
+      peers: [],
+      searchValue: "",
+      searchResults: {
+        groups: [],
+        ships: []
+      },
+      inviteError: false
+    }
+    this.search = this.search.bind(this);
+  }
+
+  componentDidMount() {
+    this.peerUpdate();
+  }
+
+  componentDidUpdate(prevProps) {
+    if (prevProps !== this.props) {
+      this.peerUpdate();
+    }
+  }
+
+  peerUpdate() {
+    let groups = Array.from(Object.keys(this.props.groups));
+    groups = groups.filter(e => !e.startsWith("/~/"));
+
+    let peers = [],
+    peerSet = new Set();
+    Object.keys(this.props.groups).map(group => {
+      peerSet.add(...this.props.groups[group]);
+    });
+    peers = Array.from(peerSet);
+
+    this.setState({ groups: groups, peers: peers });
+  }
+
+  search(event) {
+    let searchTerm = event.target.value.toLowerCase().replace("~", "");
+
+    this.setState({searchValue: event.target.value});
+
+    if (searchTerm.length < 2) {
+      this.setState({searchResults: { groups: [], ships: [] }})
+    }
+
+    if (searchTerm.length > 2) {
+      if (this.state.inviteError === true) {
+        this.setState({inviteError: false});
+      }
+
+      let groupMatches = this.state.groups.filter(e => {
+        return e.includes(searchTerm);
+      });
+      let shipMatches = this.state.peers.filter(e => {
+        return e.includes(searchTerm) && !this.props.invites.ships.includes(e);
+      });
+      this.setState({
+        searchResults: { groups: groupMatches, ships: shipMatches }
+      });
+    }
+
+    if (event.target.value.includes(",")) {
+      let isValid = true;
+      let addedShip = searchTerm.replace(",", "").trim();
+      if (!urbitOb.isValidPatp("~" + addedShip)) {
+        isValid = false;
+      }
+      if (!isValid) {
+        this.setState({ inviteError: true, searchValue: "" });
+      } else if (isValid) {
+        this.setInvite("addShip", addedShip);
+        this.setState({searchValue: ""});
+      }
+    }
+  }
+
+  setInvite(type, value) {
```

Comment:
I really dislike this pattern of using switch statements. Why not just make these into different functions??

Reply:
Sometimes things becomes habits unless someone stops you. Will amend for Monday and re-request review.